compiler/noirc_frontend/src/hir/resolution/resolver.rs

```diff
@@ -96,6 +96,9 @@ pub struct Resolver<'a> {
     /// Used to link items to their dependencies in the dependency graph
     current_item: Option<DependencyId>,
 
+    /// In-resolution names
+    resolving_ids: BTreeSet<StructId>,
```

Comment:
In what cases are we resolving multiple struct ids at once?
I'm thinking this should be a `Option<StructId>` instead. Possibly with an assert that it is None when we set it just to ensure we're only resolving 1 at a time.

Reply:
```rust
fn foo(x: impl Foo, y: impl Bar, ..) {
```

Reply:
Was that snippet meant for your other PR? I don't see what impl Traits have to do with resolving struct definitions

Reply:
Ok, that makes sense to me. When I originally was planning this and using `Ident`, I thought that there might be cases aside from structs that would benefit, but now that it's only `StructId` I think you're right. Will update 👍

Reply:
We can probably also use something similar for self-referential globals in the future like `global FOO = FOO;` but I want to avoid over designing until it is needed.

Reply:
I implemented the version with `Option`, but found that we resolve multiple structs at once when we declare them in reverse order of their dependencies:
```rust
    // Curve specification
    struct Curve { // Twisted Edwards curve
        // Coefficients in defining equation ax^2 + y^2 = 1 + dx^2y^2
        a: Field,
        d: Field,
        // Generator as point in Cartesian coordinates
        gen: Point
    }
    // Point in Cartesian coordinates
    struct Point {
        x: Field,
        y: Field
    }
```

Reply:
Can you add a similar note in the code where the BTreeSet is declared? With a simplified example